engine/src/main/java/io/zeebe/engine/processing/streamprocessor/ReProcessingStateMachine.java

```diff
@@ -216,6 +228,22 @@ private long scanLog(final long snapshotPosition) {
             lastFollowUpEventPosition = currentPosition;
           }
         }
+
+        // remember the highest key on the stream to restore the key generator after replay
+        highestRecordKey = Math.max(newEvent.getKey(), highestRecordKey);
+
+        // store the highest key of a processed command for supporting legacy processors
+        final UnifiedRecordValue recordValue =
+            recordValues.readRecordValue(newEvent, metadata.getValueType());
+        typedEvent.wrap(newEvent, metadata, recordValue);
+
+        if (MigratedStreamProcessors.isMigrated(typedEvent)
+            && typedEvent.getRecordType() == RecordType.COMMAND) {
+          followUpKeyBySourceCommandPosition.put(currentPosition, -1L);
+        }
+
+        followUpKeyBySourceCommandPosition.computeIfPresent(
+            sourceEventPosition, (position, recordKey) -> Math.max(newEvent.getKey(), recordKey));
```

Comment:
Yes. But I would postpone this because the whole class needs a refactoring. After we migrated all processors, we can remove some parts and improve the class holistically.